### PR TITLE
Kube apply button

### LINKED ISSUE #5661: Ability to apply a Kube YAML file to the cluster
### Is your enhancement related to a problem? Please describe

We have a Play Kubernetes YAML button in the Containers and Pods views, but nothing similar in the Kubernetes nav pages.

### Describe the solution you'd like

When you're in the new Kubernetes pages it would be really handy to have a button to quickly select a YAML file and apply it to the cluster.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by deboer-tim**
### What does this PR do?

Based on PR #5849, only /renderer files (latest commit) is changed here.

Creates an 'Apply YAML' button that prompts for a file, applies it to
the current context, and reports how many resources have been created.

I've added the button to the Deployments, Services, and Ingress/Routes
pages, as these are the most useful pages for a Kubernetes user. I
didn't add it to the Pods page b/c this is less likely used for a
Kube user and it might create confusion with the 'Play Kubernetes YAML'
button.

Created a SolidKubeIcon identical to how we created a SolidPodIcon
for the 'create pod from containers' action.

### Screenshot / video of UI

Screenshot 2024-02-14 at 4 16 49 PM

### What issues does this PR fix or reference?

Fixes #5661.

### How to test this PR?

Unit tests included. Pick your favourite yaml and apply!

### CODE REVIEW

**Comment by deboer-tim** [comment]
Rebased on latest Kube apply branch, updated to use new Kube icon (same as nav), and added tests. Ready for review.

File: packages/renderer/src/lib/kube/KubeApplyYAMLButton.svelte
Comment (cdrage) [inline packages/renderer/src/lib/kube/KubeApplyYAMLButton.svelte]: Could we say something like: 'Successfully applied {objects.length} resource(s)'?

Just to indicate it's successful :)

File: packages/renderer/src/lib/kube/KubeApplyYAMLButton.svelte
Comment (cdrage) [inline packages/renderer/src/lib/kube/KubeApplyYAMLButton.svelte]:
```suggestion
      message: 'Could not apply Kubernetes YAML: ' + error,
```

**Review by cdrage — CHANGES_REQUESTED** [review]
Success is working, but for some reason not erroring out anything unvalidated / working correctly, and I'm unsure if this is something we didn't check more thoroughly in the other PR:

Working:
https://github.com/containers/podman-desktop/assets/6422176/e22cbdbf-c998-47a0-accf-bae9c396b4c6


When using a random YAML (I just used `helloworld123) in a `test.yaml`):


https://github.com/containers/podman-desktop/assets/6422176/473a9954-e121-4ccc-864b-7fd6eafd2330

File: packages/renderer/src/lib/kube/KubeApplyYAMLButton.svelte
Comment (deboer-tim) [reply packages/renderer/src/lib/kube/KubeApplyYAMLButton.svelte]: Fixed.

File: packages/renderer/src/lib/kube/KubeApplyYAMLButton.svelte
Comment (deboer-tim) [reply packages/renderer/src/lib/kube/KubeApplyYAMLButton.svelte]: Fixed.

**Comment by deboer-tim** [comment]
Other comments fixed, and rebased on main.

> Success is working, but for some reason not erroring out anything unvalidated / working correctly, and I'm unsure if this is something we didn't check more thoroughly in the other PR:
> 
> When using a random YAML (I just used `helloworld123) in a `test.yaml`):

I debugged what's happening here: loadManifestsFromFile() loads and parses the file as a JSON object `helloworld123)`, then syncResources() filters it out (b/c it's not a valid manifest with a kind) and doesn't attempt to apply, and returns 0 resources pushed. Even if had tried it would have failed and returned 0 too.

This is kind of an odd case because the caller doesn't know how many valid resource(s) are supposed to be in the file to begin with, i.e. the file could have 5 resources and only 4 were valid/successfully applied, or 3 and 0. I'm thinking there should be two fixes to handle different cases:

1. applyResourcesFromFile() could throw an error if there are no valid resources in the file.
2. The button should special-case 0 and give a warning message that nothing was applied.

I'll do this in my next commit.

**Comment by deboer-tim** [comment]
New commit added, as above. Please verify.

File: packages/main/src/plugin/kubernetes-client.ts
Comment (cdrage) [inline packages/main/src/plugin/kubernetes-client.ts]:
```suggestion
```

Just need to remove this

**Comment by deboer-tim** [comment]
Looks like I didn't actually push last week, tests fixed now. @cdrage, please re-review/see if the prior comment works for you.

**Comment by cdrage** [comment]
https://github.com/containers/podman-desktop/assets/6422176/1af30bf3-c834-4c45-b497-8f85a75fcf86

Awesome work! LGTM! Tested via a faulty YAML as well as a valid one.

**Review by cdrage — APPROVED** [review]

**Review by lstocchi — COMMENTED** [review]
I see this error everytime i try to apply a yaml file (no matter if it's valid or not). 
I'm connected to the sandbox cluster.

image

**Comment by cdrage** [comment]
I've opened up https://github.com/containers/podman-desktop/pull/6086

I can confirm that this PR patched on this one, will result on correct error outputs.

**Review by lstocchi — COMMENTED** [review]
The patch made by @cdrage allowed me to read the error.
The problem is that it tries to push the resource to the default namespace even if i'm connected to lstocchi-dev.

image

image

**Comment by cdrage** [comment]
> The patch made by @cdrage allowed me to read the error.
> 
> The problem is that it tries to push the resource to the default namespace even if i'm connected to lstocchi-dev.
> 
> 
> 
> image
> 
> 
> 
> image
> 
> 

Does your test YAML has namespace default in it?

If so, that's a normal error. 

If not, then yes it should be applying it to the current selected namespace.

**Comment by deboer-tim** [comment]
I was going to say the same thing. Maybe you could share the yaml (offline)? If metadata/namespace exists and points to 'default' then this is expected behaviour, otherwise the bug is not directly in this PR but the underlying API (#5849) and I will fix.

**Comment by lstocchi** [comment]
This is the yaml
```
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  name: i-17
  annotations:
    nginx.ingress.kubernetes.io/rewrite-target: /
spec:
  ingressClassName: nginx-example
  rules:
  - http:
      paths:
      - path: /testpath1
        pathType: Prefix
        backend:
          resource:
            name: test
            kind: 'sink'
```

**Comment by deboer-tim** [comment]
My bad - underlying function supports applying namespace now but I never exposed or hooked it up to the button. Rebased on top of latest to get Charlie's error fix, exposed and added namespace, and updated tests. Working well for me now against sandbox with YAML above.

File: packages/renderer/src/lib/kube/KubeApplyYAMLButton.svelte
Comment (cdrage) [inline packages/renderer/src/lib/kube/KubeApplyYAMLButton.svelte:30]: 🎉

**Review by cdrage — CHANGES_REQUESTED** [review]
Tested it 3 ways with the following YAML:

```yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    app: helloworld-pod
  name: helloworld-pod
spec:
  containers:
  - image: docker.io/cdrage/helloworld:latest
    name: helloworld
    ports:
    - containerPort: 8080
      hostPort: 8080
 ```
 
 
 1. 🎉  `namespace: kube-system` added, even though kubernetes context is `default`. Deploys to `kube-system` which is intentional and awesome.
 2. 🎉 `namespace: default` added, while being on default namespace.
 3. 🛑 No namespace: No Pod is deployed despite it saying "Resources applied". If you apply the same yaml via `kubectl apply` it works fine.

File: packages/main/src/plugin/kubernetes-client.spec.ts
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-client.spec.ts]: Same problem.

File: packages/main/src/plugin/kubernetes-client.spec.ts
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-client.spec.ts]:
```suggestion
```

File: packages/main/src/plugin/kubernetes-client.spec.ts
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-client.spec.ts]:
```suggestion
expectedError = err;
```

File: packages/main/src/plugin/kubernetes-client.spec.ts
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-client.spec.ts:1224]: This test does not fail if apply does not throw an error, but test name says it 'should'.

File: packages/main/src/plugin/kubernetes-client.spec.ts
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-client.spec.ts:1222]:
```suggestion
  let expectedError: unknown;
  try {
```

File: packages/main/src/plugin/kubernetes-client.spec.ts
Comment (dgolovin) [inline packages/main/src/plugin/kubernetes-client.spec.ts:1229]:
```suggestion
  expect(expectedError).to.be.a('Error');
  expect((err as Error).message).equal('File missing-file.yaml does not exist');
});
```

File: packages/renderer/src/lib/kube/KubeApplyYAMLButton.svelte
Comment (dgolovin) [inline packages/renderer/src/lib/kube/KubeApplyYAMLButton.svelte:50]: Better to use coercion here
```suggestion
      message: 'Could not apply Kubernetes YAML: ' + String(error),
```

**Review by dgolovin — CHANGES_REQUESTED** [review]

**Comment by cdrage** [comment]
> Tested it 3 ways with the following YAML:
> 
> ```yaml
> apiVersion: v1
> kind: Pod
> metadata:
>   labels:
>     app: helloworld-pod
>   name: helloworld-pod
> spec:
>   containers:
>   - image: docker.io/cdrage/helloworld:latest
>     name: helloworld
>     ports:
>     - containerPort: 8080
>       hostPort: 8080
> ```
> 
> 1. 🎉  `namespace: kube-system` added, even though kubernetes context is `default`. Deploys to `kube-system` which is intentional and awesome.
> 2. 🎉 `namespace: default` added, while being on default namespace.
> 3. 🛑 No namespace: No Pod is deployed despite it saying "Resources applied". If you apply the same yaml via `kubectl apply` it works fine.

Scratch that...

I found the problem and it was a lingering "Terminating" helloworld-pod that was not allowing the YAML to complete, and that's why it was giving the false "0 Resources Applied" being shown.

My "hello world" container takes too long to shut down.

Intentional feature of applying it looks!

**Review by cdrage — APPROVED** [review]
LGTM! Just updating the tests of @dgolovin 's comments and we should be good to go 🚀 🚀 🚀

**Comment by deboer-tim** [comment]
@dgolovin  tests updated, please review

**Comment by jeffmaury** [comment]
The Pod page still have Play Kubernetes YAML and not Apply YAML

**Comment by deboer-tim** [comment]
> The Pod page still have Play Kubernetes YAML and not Apply YAML

Play Kube is fundamentally different from Apply: it works against Podman (hence the name), or can target Kubernetes but does a create (i.e. won't change existing resources). Apply matches 'kubectl apply' which does 'create or patch', and only works on Kubernetes.

Personally I think we should reduce Play Kube to only work on Podman and use Apply for Kube, but this is not the only area where Pods fall in both camps, having two buttons on the Pods page would be even more confusing, and this gets into some political questions. My goal here was to add Apply Kube to pages like Deployments and Services which are clearly Kube pages and it is expected, and defer/avoid disrupting/wait for feedback on the Pods page.

**Comment by deboer-tim** [comment]
Rebased to fix conflicts.

**Comment by deboer-tim** [comment]
@dgolovin another review please?

**Review by dgolovin — APPROVED** [review]
LGTM

**Review by cdrage — APPROVED** [review]
After your changes, I've tested this again against a "error out YAML" as well as a functional yaml with multiple resources.

Everything LGTM! Error-wise too.

Hitting the merge button 🎉

**Pull request merged**
